components/brave_perf_predictor/browser/third_party_extractor.cc

```diff
@@ -0,0 +1,148 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/third_party_extractor.h"
+
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/values.h"
+#include "components/grit/brave_components_resources.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "ui/base/resource/resource_bundle.h"
+
+namespace brave_perf_predictor {
+
+namespace {
+
+RE2 DOMAIN_IN_URL_REGEX(":\\/\\/(.*?)(\\/|$)");
+RE2 DOMAIN_CHARACTERS("([a-z0-9.-]+\\.[a-z0-9]+)");
+RE2 ROOT_DOMAIN_REGEX("([^.]+\\.([^.]+|(gov|com|co|ne)\\.\\w{2})$)");
+
+base::Optional<std::string> GetDomainFromOriginOrUrl(
+    const std::string& origin_or_url) {
+  std::string domain;
+  if (RE2::PartialMatch(origin_or_url, DOMAIN_IN_URL_REGEX, &domain))
+    return domain;
+  if (RE2::PartialMatch(origin_or_url, DOMAIN_CHARACTERS, &domain))
+    return domain;
+  return base::nullopt;
+}
+
+std::string GetRootDomain(const std::string& domain) {
+  std::string root_domain;
+  if (RE2::PartialMatch(domain, ROOT_DOMAIN_REGEX, &root_domain)) {
+    return root_domain;
+  } else {
+    return domain;
+  }
+}
+
+}  // namespace
+
+ThirdPartyExtractor::ThirdPartyExtractor() = default;
+
+bool ThirdPartyExtractor::InitializeFromResource() {
+  const auto resource_id = IDR_THIRD_PARTY_ENTITIES;
```

Comment:
Sadly we are not able to add trace events here, but maybe (1) add `TODO: insert trace event here` (2) add a UMA histogram timer, just to check timings locally on `brave://histograms`?